### ISSUE TITLE
Requesting metadata proposal

### ISSUE THREAD

**Issue by henryiii** [scikit-build/dynamic-metadata#21]
Some hooks may want to access other's metadata. I think this could be supported in three stages:

First, the hooks could be redesigned to modify an existing dictionary of metadata in place. The new hook signature would be:

```python
def dynamic_metadata(
    metadata: MutableMapping[str, str | dict[str, str | None] | list[str]],
    field: str,
    settings: Mapping[str, object] | None = None,
) -> None:
    "Modify the metadata inplace"
```

The contents of the `metadata` dict would include static metadata, and the hook is expected to replace or add `field`. At this stage, nothing is stated about the order of the hooks run.

Second stage could be to require `version` be computed before other hooks. I believe this is the current implantation in hatchling, and is the reason for adding this (hatch-fancy-pypi-readme requires the version pre-computed).

The third stage would be to add topological sorting. This would probably be easier to add after Python 3.9 becomes the minimum required version for most backends. A new hook would be added:

```python
def metadata_reqiures(
    field: str,
    settings: Mapping[str, object] | None = None,
) -> list[str]:
    "Return fields needing to be computed first"
```

And backends should topologically sort before running.

**Comment by henryiii** [comment]
Anyone can read the pyproject.toml. This is specifically the metadata, since it could be dynamic, and therefore you could not just read it from the file.

**Comment by LecrisUT** [comment]
But the main idea of this is to edit the `metadata` right? So would the edits (and reads) be on the `project` table or other tables in `pyproject.toml` like the `tool.*`? I suspect it's the former right?

**Comment by henryiii** [comment]
The metadata is build using the `[project]` table, though anything listed in the `project.dynamic` field can come from other places. If a user lists more than one field in `project.dynamic`, then can one field see what another one produces? The use case was for `readme`. If you have a dynamic `readme`, it might want to be able to access a dynamic `version`. `hatch-fancy-pypi-readme` allows you to write something like `${version}` in your readme and it will inject the version there.

The metadata is the stuff that goes into PKG-INFO.

**Comment by henryiii** [comment]
I ran into another use case, and one that's pretty important to me. For pybind11, the `pybind11` package needs to have a `[global]` optional-dependency, and that needs to depend on `pybind11-global==${version}`, where that's the current version.

Maybe it would be better to have an input dict, where anything in `metadata_reqiures` is guaranteed to be present, instead of allowing arbitrary mutation.

**Comment by LecrisUT** [comment]
> Maybe it would be better to have an input dict, where anything in `metadata_reqiures` is guaranteed to be present, instead of allowing arbitrary mutation.

Agreed. Related to the `asyncio.Future` comment, i.e. what are the unevaluated fields filled with in the meantime. Probably could do it in other ways also, e.g. constructing the `__getitem__` to add all of the dynamic metadata hooks before evaluating.

Also these metadata are only calculated once and saved in `PKG-INFO` or equivalent and any of the `dynamic_metadata` is no longer evaluated right? It's `pip` responsibility to figure out where/how to get that metadata?

**Comment by henryiii** [comment]
They are not present. I'm trying to put together a PR for scikit-build-core, maybe it will be clearer in code. I'm going with your the non-mutating idea for starters. I don't think you can... Oh, you mean to make the whole thing async? If it doesn't actually need to be async, I think it's probably best not, you might have to start up a async loop just for this, and starting one if one is running already is an issue, etc. Now that 3.9 is the oldest non-EoL, making a proposal around 3.9+ isn't an issue (and we can use just the backport for 3.8).

I'll try to have something to show soon.
